### PR TITLE
Remove leaf bias from BTreeMap's BoxedNode and NodeRef

### PR DESCRIPTION
**Pull request by ssomers**
Improve readability - there is no genuine problem here, after #56648.

`BoxedNode` and `NodeRef` contain a pointer to a node that is typed as `LeafNode`. In case of EMPTY_ROOT_NODE, that pointer may not be properly aligned and does not point to anything matching `LeafNode<K, V>`. Only `NodeRef`'s documentation mentions this. 

It seemed more logical to me to give these pointers as type the least common denominator, `NodeHeader<K, V>` (which only exists since  #56648). That simplifies the simplest access `as_header` and levels the playing field for leaf and internal nodes.

One could argue that most nodes are in fact leaves, and most access to nodes is to their leaf-portion, keys and values. Then I'd still like to fix the documentation of `BoxedNode`.

### CODE REVIEW

**Comment by ssomers** [comment]
One could assign this to @rkruppe since it was touched upon at the end of #67812

File: src/liballoc/collections/btree/node.rs
Comment (rkruppe) [inline src/liballoc/collections/btree/node.rs]: The shared root is neither boxed nor uniquely owned by one `BoxedNode`, so although it's good to note that this pointer can refer to the shared root, wording it like this is misleading IMO. I'd suggest the following outline instead:

- This is usually an owned pointer to a LeafNode/InternalNode, but it can also be a pointer (no ownership) to EMPTY_ROOT_NODE
- No destructor in part because we don't know which of these variants it is
- All of these types have a `NodeHeader<K, V>` prefix (even `EMPTY_ROOT: NodeHeader<(), ()>`), so that's the pointee type we store

File: src/liballoc/collections/btree/node.rs
Comment (rkruppe) [inline src/liballoc/collections/btree/node.rs]: I see that this is consistent with `from_internal` but I'm tempted to suggest rewriting both into `Box::into_unique(node).cast()`  to get rid of the gratuitous `unsafe`.

File: src/liballoc/collections/btree/node.rs
Comment (rkruppe) [inline src/liballoc/collections/btree/node.rs]: Creating a reference here risks UB and is also completely pointless. Just use `self.node.as_ptr() as *mut LeafNode<K, V>`.

File: src/liballoc/collections/btree/node.rs
Comment (rkruppe) [inline src/liballoc/collections/btree/node.rs]: Nit: can now be simplified to `NonNull::from(&EMPTY_ROOT_NODE)`.

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]
Some nits, and more importantly, one safety issue.

File: src/liballoc/collections/btree/node.rs
Comment (ssomers) [reply src/liballoc/collections/btree/node.rs]: I'm confused by and can't find a description of "prefix" in a pointer context. Does that mean the data it points at always has at least the size and layout of `NodeHeader<K, V>`? Not that I know any better word for it. Does "prefix" also imply that pointer alignment is at least as high as that of `NodeHeader<K, V>`?

File: src/liballoc/collections/btree/node.rs
Comment (ssomers) [reply src/liballoc/collections/btree/node.rs]: The compiler and I think you're wrong there: something must be done to transmute the `NodeHeader<(), ()>` to `NodeHeader<K, V>`. Why we also have to tag it as mutable, beats me.

Earlier, I got off on a tangent to adjust `BoxedNode` to this actual type of `EMPTY_ROOT_NODE`, which implies removing K and V, or pretending to understand PhantomData, but I managed to squeeze that monster back into the box before it ate me.

**Comment by ssomers** [comment]
Meanwhile, I managed to reproduce and hopefully fix the failing pr build. Awaiting a long build.

File: src/liballoc/collections/btree/node.rs
Comment (ssomers) [reply src/liballoc/collections/btree/node.rs]: Done, and alloc tests and Miri are happy

File: src/liballoc/collections/btree/node.rs
Comment (ssomers) [reply src/liballoc/collections/btree/node.rs]: Done, and alloc tests and Miri are happy

File: src/liballoc/collections/btree/node.rs
Comment (rkruppe) [reply src/liballoc/collections/btree/node.rs]: To be precise, I mean that all of `InternalNode<K, V>`, `LeafNode<K, V>` and `NodeHeader<(), ()>`:

- are at least as large as `NodeHeader<K, V>`
- have alignment at least as large as `NodeHeader<K, V>`
- store the same kinds of data, at the same offsets, as a `NodeHeader<K, V>` in the first `size_of::<NodeHeader<K, V>>`

This is the reason why `as_header()` is unconditionally safe. Alignment in particular is probably worth calling out separately, as it's not really implied by "prefix" (e.g., you might say `i32` is a prefix of `#[repr(C)] struct { a: i32, b: i64 }` but alignment differs).

File: src/liballoc/collections/btree/node.rs
Comment (rkruppe) [reply src/liballoc/collections/btree/node.rs]: I forgot about the type parameters, mea culpa. `NonNull::from(&EMPTY_ROOT_NODE).cast()` should work, though.

File: src/etc/gdb_rust_pretty_printing.py
Comment (ssomers) [inline src/etc/gdb_rust_pretty_printing.py]: What a kludge... #67686 would avoid that part.

File: src/etc/gdb_rust_pretty_printing.py
Comment (ssomers) [inline src/etc/gdb_rust_pretty_printing.py]: This `if` and moving `length` may not have been necessary. I don't quite understand yet how it's called.

File: src/etc/gdb_rust_pretty_printing.py
Comment (ssomers) [inline src/etc/gdb_rust_pretty_printing.py]: The missing `,1 ` in `replace` is definitely a bug in rust-gdb nightly, but I'm not planning on making an issue for it.

**Comment by rkruppe** [comment]
I know nothing about the gdb pretty printing, so I would prefer if the changes to it were kept to the bare minimum required by the other changes in this PR, or alternatively, if someone who knows anything about the pretty printers reviewed those parts.

**Comment by ssomers** [comment]
I'll first submit a PR fixing and testing pretty printing as is now, and they can judge whether the near-invalid memory access is cool or cripple.

File: src/etc/gdb_rust_pretty_printing.py
Comment (ssomers) [reply src/etc/gdb_rust_pretty_printing.py]: moved off to #68098

File: src/etc/gdb_rust_pretty_printing.py
Comment (ssomers) [reply src/etc/gdb_rust_pretty_printing.py]: moved off to #68098

**Comment by JohnCSimon** [comment]
Ping from triage: 
@ssomers - Can you please post your status and address the change request from **rkruppe** ? t

**Comment by ssomers** [comment]
This PR is now blocked by #68098 and won't automatically merge after it. As far as I know, I've addressed all change requests.

**Comment by ssomers** [comment]
Nothing to see here, folks. I just played around in the GitHub interface and rebased on the #68098 commit.

File: src/etc/gdb_rust_pretty_printing.py
Comment (ssomers) [reply src/etc/gdb_rust_pretty_printing.py]: #67686 is merged so this looks better now

**Comment by JohnCSimon** [comment]
> This PR is now blocked by #68098 and won't automatically merge after it (or will need changes). As far as I know, I've addressed all change requests.

@rkruppe - can you please review this pr?

**Comment by ssomers** [comment]
For clarity, this PR is still blocks by #68098, and is rebased on it so might merge automatically after it. As far as I know, we have reviewed and addressed everything apart from the part dependent on #68098.

**Comment by rkruppe** [comment]
Yes, AFAIK there's nothing to be done here before #68098 is merged and once that's done this PR can be rebased and most likely merged (NB: the first commit here has a different hash than the commit in that PR, so some manual `git rebase --onto` will be needed).

**Comment by ssomers** [comment]
Well spotted, I think that before rebasing _this one_ on _that one_, I rebased _that one_ on _master_ and forgot to push.

**Comment by ssomers** [comment]
Closing for now (and it looks to be a very long time) because being based on another PR's commit misguides GitHub.

**Pull request closed without merging**
